### ISSUE TITLE
I can't generate an RSA-signed tab - KeyRejected("InvalidEncoding")

### ISSUE THREAD

**Issue by george-cosma** [tock/elf2tab#77]
Hello! I was recently experimenting with elf2tab and I saw that I could sign an app using RSA4096. I tried following the tutorial on the README, but I can't get it working. I did this on the c_hello example from libtock-c.

Error:
```
thread 'main' panicked at 'RSA4096 could not be parsed: KeyRejected("InvalidEncoding")', /home/<user>/.cargo/registry/src/github.com-1ecc6299db9ec823/elf2tab-0.11.0/src/convert.rs:998:17
stack backtrace:
   0: rust_begin_unwind
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panicking.rs:579:5
   1: core::panicking::panic_fmt
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/panicking.rs:64:14
   2: elf2tab::convert::elf_to_tbf
   3: elf2tab::main
```
Commands executed:
```bash
$ openssl genrsa -aes256 -out tockkey.private.pem 4096
$ openssl rsa -in tockkey.private.pem -outform der -out tockkey.private.der
$ openssl rsa -in tockkey.private.pem -outform der -pubout -out tockkey.public.der
```

```bash
$ openssl version
OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
```

```bash
elf2tab --rsa4096-private tockkey.private.der --rsa4096-public tockkey.public.der -n c_hello --stack 2048 --app-heap 1024 --kernel-heap 1024 --kernel-major 2 --kernel-minor 0 -v -o build/c_hello.tab build/cortex-m0/cortex-m0.elf build/cortex-m3/cortex-m3.elf build/cortex-m4/cortex-m4.elf build/cortex-m7/cortex-m7.elf
# basically what `make` executes inside libtock-c, just with the rsa parameters added.
```

Just for context, here's where I think the issue is occuring: [let key_pair = signature::RsaKeyPair::from_der(&private_key_der)](https://github.com/tock/elf2tab/blob/bf5f46071e56642cd8d73e2bc84f5bc2582a212b/src/convert.rs#L996-L1000)

**Comment by bradjc** [comment]
Interesting. I just copied these steps and they worked for me.

```
elf2tab --rsa4096-private isstest.private.der --rsa4096-public isstest.public.der -n c_hello --stack 2048 --app-heap 1024 --kernel-heap 1024 --kernel-major 2 --kernel-minor 0 -v -o build/c_hello.tab build/cortex-m4/cortex-m4.elf
Creating "build/cortex-m4/cortex-m4.tbf"
Min RAM size from segments in ELF: 516 bytes
Number of writeable flash regions: 0
Kernel version: 2.0
  Adding segment. Offset: 76 (0x4c). Length: 880 (0x370) bytes.
    Contains section .crt0_header. Offset: 76 (0x4c). Length: 40 (0x28) bytes.
    Contains section .text. Offset: 116 (0x74). Length: 840 (0x348) bytes.
  Adding segment. Offset: 956 (0x3bc). Length: 168 (0xa8) bytes.
    Contains section .got. Offset: 956 (0x3bc). Length: 40 (0x28) bytes.
    Contains section .data. Offset: 996 (0x3e4). Length: 128 (0x80) bytes.
      Including relocation data (.rel.data). Length: 40 (0x28) bytes.
    Contains section .bss. Offset: 1124 (0x464). Length: 348 (0x15c) bytes.
TBF Header:
               version:        2        0x2
           header_size:       76       0x4C
            total_size:     4096     0x1000
                 flags:        1        0x1

        init_fn_offset:       41       0x29
        protected_size:        0        0x0
      minimum_ram_size:     4612     0x1204

        init_fn_offset:       41       0x29
        protected_size:        0        0x0
      minimum_ram_size:     4612     0x1204
     binary_end_offset:     1168      0x490
           app_version:        0        0x0

    kernel version: ^2.0
Added PKCS#1v1.5 RSA4096 signature credential.
```

I'm on mac:

```
$ openssl version
OpenSSL 1.1.1t  7 Feb 2023
```

I wonder what is different about your .der file?

Maybe try:

```
$ openssl rsa -in tockkey.private.der -inform der -outform text -text -noout
```

I get something like:
```
RSA Private-Key: (4096 bit, 2 primes)
modulus:
    00:e7:3f:f6:71:93:3e:62:df:5c:00:95:26:ac:4b:
    4c:bf:38:31:30:2f:88:85:64:56:ea:3e:d9:e0:fd:
    48:0f:d3:f8:2d:35:82:67:6c:e1:df:8f:09:0d:64:
...
```

**Comment by george-cosma** [comment]
Yhea, something very strange is happening.

The command you told me about seems to output something similar to what you have
```bash
$ openssl rsa -in tockkey.private.der -inform der -outform text -text -noout | head 
RSA Private-Key: (4096 bit, 2 primes)
modulus:
    00:b8:fe:63:68:8d:5c:3c:24:47:5c:27:46:5d:c5:
    83:2b:86:05:ef:1d:2e:74:3d:07:80:84:76:59:7d:
    49:d3:5a:72:3c:70:bf:d5:47:97:5b:52:a9:61:40:
```

Initially, I tried did everything on an Ubuntu virtual machine, then I tried the exact same steps (minus compiling the .elf file) on WSL, and then the same thing on Windows. **On windows, it worked perfectly only if I generated the key using windows!** That is to say, running elf2tab with the WSL-generated keys didn't work, but after I regenerated them on Windows it worked.

I suspect something is going on with the `.der` files, possibly openssl?

```bash
# OpenSSL version on Windows
$ openssl version
OpenSSL 1.1.1s  1 Nov 2022
```

**Comment by bradjc** [comment]
Hmmm...you can try #78? We're just outsourcing the key reading to a library, and it's possible that something about the format doesn't work with the library. Also, I'm not really sure how out of date the ring library we are using is (there doesn't seem to be much activity, but it's two years old).

Or we have to be more specific about what type of openssl can generate usable keys...

**Comment by george-cosma** [comment]
I have noticed a small difference in the way linux and windows version of openssl emit the `der` files:
```bash
# Windows version
$ xxd tockkey.private.der | head 
00000000: 3082 092a 0201 0002 8202 0100 c4af 372e  0..*..........7.
00000010: 8914 242d 2ab5 479e e689 157c f50c f824  ..$-*.G....|...$
00000020: af6c f4b4 b976 df9a d372 5748 a012 0780  .l...v...rWH....
00000030: 5f7d a041 973d 07d9 ff42 a6d0 4c95 4f8f  _}.A.=...B..L.O.
00000040: ec67 24d1 2bfb 4e6d 6aea 9d8a 4a87 26c0  .g$.+.Nmj...J.&.
00000050: 8a4d 4a45 9703 f0d7 f462 0939 b196 6964  .MJE.....b.9..id
00000060: a580 c496 097a 22ce 456b e1d1 7630 29fd  .....z".Ek..v0).
00000070: 9be6 8452 0659 e7f0 ce6e 292f 023c 63a5  ...R.Y...n)/.<c.
00000080: 556b fd4c b4e2 1faa 7294 c44c 020c b0a8  Uk.L....r..L....
00000090: d880 666a 57e5 5cf9 b2f2 ad23 3fbe 2c65  ..fjW.\....#?.,e

# WSL Version
$ xxd tockkey.private.wsl.der | head
00000000: 3082 0944 0201 0030 0d06 092a 8648 86f7  0..D...0...*.H..
00000010: 0d01 0101 0500 0482 092e 
# Notice here how this is now all the same with the windows version:
                                   3082 092a 0201  ..........0..*..
00000020: 0002 8202 0100 c4af 372e 8914 242d 2ab5  ........7...$-*.
00000030: 479e e689 157c f50c f824 af6c f4b4 b976  G....|...$.l...v
00000040: df9a d372 5748 a012 0780 5f7d a041 973d  ...rWH...._}.A.=
00000050: 07d9 ff42 a6d0 4c95 4f8f ec67 24d1 2bfb  ...B..L.O..g$.+.
00000060: 4e6d 6aea 9d8a 4a87 26c0 8a4d 4a45 9703  Nmj...J.&..MJE..
00000070: f0d7 f462 0939 b196 6964 a580 c496 097a  ...b.9..id.....z
00000080: 22ce 456b e1d1 7630 29fd 9be6 8452 0659  ".Ek..v0)....R.Y
00000090: e7f0 ce6e 292f 023c 63a5 556b fd4c b4e2  ...n)/.<c.Uk.L..
```

The WSL/Linux version has an additional header. Perhaps it could be omitted using a parameter to openssl?

**Comment by george-cosma** [comment]
Yhea, the PK8 works wonderfully, no issues (minus `ring` not compiling on Windows, but with a fix to that it works)

**Comment by bradjc** [comment]
Ok then it seems like pk8 is the way to go. It is also recommended in ring (probably for this reason...).

**Comment by bradjc** [comment]
Seems like #78 is what we want. Note, the only thing we can currently do with rsa signatures is view them with tockloader so I'm not worried about changing interfaces or maintaining any sort of stability.
